uniffi_bindgen/src/interface/mod.rs

```diff
@@ -119,12 +125,62 @@ impl<'ci> ComponentInterface {
         self.errors.to_vec()
     }
 
+    /// Calculate a numeric checksum for this ComponentInterface.
+    ///
+    /// The checksum can be used to guard against accidentally using foreign-language bindings
+    /// generated from one version of an interface with the compiled Rust code from a different
+    /// version of that interface. It offers the following properties:
+    ///
+    ///   - Two ComponentIntefaces generated from the same WebIDL file, using the same version of uniffi
+    ///     and the same version of Rust, will always have the same checksum value.
+    ///   - Two ComponentInterfaces will, with high probability, have different checksum values if:
+    ///         - They were generated from two different WebIDL files.
+    ///         - They were generated by two different versions of uniffi
+    ///
+    /// The checksum may or may not change depending on the version of Rust used; since we expect
+    /// consumers to be using the same executable to generate both the scaffolding and the bindings,
+    /// assuming the same version of Rust seems acceptable.
+    ///
+    /// Note that this is designed to prevent accidents, not attacks, so there is no need for the
+    /// checksum to be cryptographically secure.
+    ///
+    /// TODO: it's not clear to be if the derivation of `Hash` is actually deterministic enough to
+    /// ensure the guarantees above, or if it might be sensitive to e.g. compiler-driven re-ordering
+    /// of struct field. Let's see how it goes...
+    pub fn checksum(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        // Our implementation of `Hash` mixes in all of the public API of the component,
+        // as well as the version string of uniffi.
```

Comment:
You could imagine a version of this that just does `sha256(uniffi_version + webidl_file_contents)` for a similar effect, rather than (ab)using the `DefaultHasher` like this. I've opted for this approach because I think it's less committal about our ongoing use of WebIDL post-MVP.